docs/pipeline-yml-style.md

```diff
@@ -0,0 +1,97 @@
+# Azure pipeline YAML style
+
+Our pipeline implementation is shaped by a few AzDO pipeline and YAML quirks. This doc contains notes on these quirks and how they influence how we've written the pipeline YAML.
+
+These notes also apply to the [microsoft/go](https://github.com/microsoft/go) pipelines.
+
+Useful AzDO Pipeline YAML docs:
+* [YAML schema reference for Azure Pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema)
+  * [Shortcut steps: `pwsh`, `script`, `publish`, etc.](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/steps?view=azure-pipelines)
+* [Template types & usage](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=azure-devops)
+* [Build and release tasks](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/?view=azure-devops)
+
+## One stage per job
+
+If a job fails in a running AzDO build with only one stage, you must wait for every single job to fail before you can retry the failed job. This is a problem: if a long job fails early but other long jobs are still running, we waste time waiting for the whole stage to finish.
```

Comment:
I'm not sure I understand this, it seems to contradict itself. You say on stage per job is bad, then say the workaround is to have one stage per job?

Reply:
Reading this again, nasty ambiguity here:

> If a job fails in a running AzDO build with only one stage ...

With what I had in mind at the time, it means this:

> If a job [out of many other jobs] fails in a running AzDO build [and the build has] only one stage [that contains all the jobs] ...

But could also mean:

> If a job fails in a running AzDO build [and that job has] one stage ...

I think I'll replace this with an example to cut out the English. 😄

Reply:
Ahh ok, I understand now. Possibly that was made worse by me not understanding exactly what a job vs a stage is. The replacement explanation is even better though :)